src/v/ssx/this_fiber_id.h

```diff
@@ -0,0 +1,28 @@
+#include "seastarx.h"
+#include "vassert.h"
+
+#include <seastar/core/reactor.hh>
+
+namespace ssx {
+
+/// Return unique id of the current fiber
+///
+/// Terminate will be called if the function is called outside of the coroutine.
+/// If one coroutine awaits another one they both will have the same fiber id.
+inline uint64_t this_fiber_id() {
+    auto task = ss::engine().current_task();
+    vassert(task != nullptr, "this_fiber_id can only be called from coroutine");
+    while (true) {
+        auto child = task->waiting_task();
+        if (child == nullptr) {
+            break;
+        } else {
+            task = child;
+        }
+    }
+    uint64_t task_id = 0;
+    std::memcpy(&task_id, &task, sizeof(task_id));
```

Comment:
Am I right in that `task` is a `task*` not an integer, so how does this work?

Reply:
Yes. It's a type punning. The top level task is the one which the coffee uses to identify a fiber. So casting that pointer to int gives a unique integer id.